docs/topics/jvm/java-to-kotlin-nullability-guide.md

```diff
@@ -0,0 +1,328 @@
+numbers.add(1);
+numbers.add(2);
+
+System.out.println(numbers.get(0));
+//numbers.get(5) // Exception!
+```
+{id="functions-returning-null-java"}
+
+The Kotlin standard library often provides functions whose names indicate whether they can possibly return a `null` value.
+This is especially common in the collections API:
+
+```kotlin
+// Kotlin
+fun main() {
+//sampleStart
+    // The same code as in Java:
+    val numbers = listOf(1, 2)
+    
+    println(numbers[0])  // Can throw IndexOutOfBoundsException if the collection is empty
+    //numbers.get(5)     // Exception!
+
+    // More abilities:
+    println(numbers.firstOrNull())
+    println(numbers.getOrNull(5)) // null
+//sampleEnd
+}
+```
+{kotlin-runnable="true" id="functions-returning-null-kotlin"}
+
+## Aggregate operations
+
+When you need to get the biggest element or `null` if there are no elements, in Java you would use
+the [Stream API](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/package-summary.html):
+
+```java
+// Java
+var numbers = new ArrayList<Integer>();
+var max = numbers.stream().max(Comparator.naturalOrder()).orElse(null);
+System.out.println("Max: " + max);
+```
+{id="aggregate-functions-java"}
+
+In Kotlin, use [aggregate operations](collection-aggregate.md):
+
+```kotlin
+// Kotlin
+fun main() {
+//sampleStart
+    val numbers = listOf<Int>()
+    println("Max: ${numbers.maxOrNull()}")
+//sampleEnd    
+}
+```
+{id="aggregate-functions-kotlin"}
+
+Learn more about [collections in Java and Kotlin](java-to-kotlin-collections-guide.md).
+
+## Casting types safely
+
+When you need to safely cast a type, in Java you would use the `instanceof` operator and then check how well it worked:
+
+```java
+// Java
+int getStringLength(Object y) {
+    return y instanceof String x ? x.length() : -1;
+}
+
+void main() {
+    System.out.println(getStringLength(1)); // Prints `-1`
+}
+```
+{id="casting-types-java"}
+
+To avoid exceptions in Kotlin, use the [safe cast operator](typecasts.md#safe-nullable-cast-operator) `as?`, which returns `null` on failure:
+
+```kotlin
+// Kotlin
+fun main() {
+    println(getStringLength(1)) // Prints `-1`
+}
+
+fun getStringLength(y: Any): Int {
+    val x: String? = y as? String // null
+    return x?.length ?: -1 // Returns -1 because `x` is null
+}
+```
+{kotlin-runnable="true" id="casting-types-kotlin"}
+
+> In the Java example above, the function `getStringLength()` returns a result of the primitive type `int`.
+To make it return `null`, you can use the ["boxed" type](https://docs.oracle.com/javase/tutorial/java/data/autoboxing.html) `Integer`.
+However, it's more resource-efficient to make such functions return a negative value and then check the value –
+you would do the check anyway, but no additional boxing is performed this way.
+>
+> {type="note"}
+
+## What's next?
+
+* Visit [Kotlin Koans](koans.md) – complete exercises to learn Kotlin syntax. Each exercise is created as a failing unit test,
```

Comment:
Too many links here.
At least we could remove Koand links: I've checked the analytics for the Collections guide for the last three months and found that there are a few of the clicks (33). What do you think?

Reply:
I agree.